core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala

```diff
@@ -367,21 +368,55 @@ object Protocol {
       !metadata.configuration.contains(DeltaConfigs.CREATE_TABLE_IGNORE_PROTOCOL_DEFAULTS.key),
       "Should not have the table property " +
         s"${DeltaConfigs.CREATE_TABLE_IGNORE_PROTOCOL_DEFAULTS.key} stored in table metadata")
-    val (readerVersion, writerVersion, enabledFeatures) =
+    val (readerVersion, writerVersion, minRequiredFeatures) =
       minProtocolComponentsFromAutomaticallyEnabledFeatures(spark, metadata)
 
     // Increment the reader and writer version to accurately add enabled legacy table features
     // either to the implicitly enabled table features or the table feature lists
     val required = Protocol(
         readerVersion.max(current.minReaderVersion), writerVersion.max(current.minWriterVersion))
-        .withFeatures(enabledFeatures)
+        .withFeatures(minRequiredFeatures)
     if (!required.canUpgradeTo(current)) {
+      // When the current protocol does not satisfy metadata requirement, some additional features
+      // must be supported by the protocol. We assert those features can actually perform the
+      // auto-update.
+      assertMetadataTableFeaturesAutomaticallySupported(
+        current.implicitlyAndExplicitlySupportedFeatures,
+        required.implicitlyAndExplicitlySupportedFeatures)
       Some(required.merge(current))
     } else {
       None
     }
   }
 
+  /**
+   * Ensure all features listed in `currentFeatures` are also listed in `requiredFeatures`, or, if
+   * one is not listed, it must be capable to auto-update a protocol.
+   *
+   * Refer to [[FeatureAutomaticallyEnabledByMetadata.automaticallyUpdateProtocolOfExistingTables]]
+   * to know more about auto-update capable features.
+   *
+   * Note: Caller must make sure `requiredFeatures` is obtained from a min protocol that satisfies
+   * a table metadata.
+   */
+  private def assertMetadataTableFeaturesAutomaticallySupported(
+      currentFeatures: Set[TableFeature],
+      requiredFeatures: Set[TableFeature]): Unit = {
+    val (autoUpdateCapableFeatures, nonAutoUpdateCapableFeatures) =
+      requiredFeatures.diff(currentFeatures)
+        .collect { case f: FeatureAutomaticallyEnabledByMetadata => f }
+        .partition(_.automaticallyUpdateProtocolOfExistingTables)
+    if (nonAutoUpdateCapableFeatures.nonEmpty) {
+      // The "current features" we give the user are which from the original protocol, plus
+      // features newly supported by table properties in the current transaction, plus
+      // metadata-enabled features that are auto-update capable. The first two are provided by
+      // `currentProtocol`.
```

Comment:
There is no `currentProtocol` locally in this function.

Reply:
Should be `currentFeatures`. Fixed!